### PR TITLE
New Resource: `azurerm_dev_center_network_connection`

### PR DESCRIPTION
**Pull request by neil-yechenwei**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description
This PR is to support new resource Dev Center Network Connection.

API Spec: https://github.com/Azure/azure-rest-api-specs/blob/09741dc7c9e43a8f9401d782186cac723dc4af71/specification/devcenter/resource-manager/Microsoft.DevCenter/stable/2023-04-01/vdi.json#L707

Overview: https://learn.microsoft.com/en-us/azure/dev-box/how-to-configure-network-connections?tabs=AzureADJoin

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [x] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
--- PASS: TestAccDevCenterNetworkConnection_basic (788.66s)
--- PASS: TestAccDevCenterNetworkConnection_requiresImport (721.97s)
--- PASS: TestAccDevCenterNetworkConnection_complete (1301.52s)
--- PASS: TestAccDevCenterNetworkConnection_update (1595.60s)
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* New Resource: `azurerm_dev_center_network_connection`

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/devcenter/dev_center_network_connection_resource_test.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+			Config: r.complete(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep("domain_password"),
+	})
+}
+
+func TestAccDevCenterNetworkConnection_update(t *testing.T) {
```
Comment (ms-zhenhua) [inline internal/services/devcenter/dev_center_network_connection_resource_test.go:66]: shall we use `basic->complete->update->basic` to test add/update/remove properties?

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hi @neil-yechenwei, Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

**Comment by neil-yechenwei** [comment]
@ms-zhenhua , Thanks for the comments. I updated PR. Please take another look. Thanks.

File: internal/services/devcenter/dev_center_network_connection_resource_test.go (hunk 1 above)
Comment (neil-yechenwei) [reply internal/services/devcenter/dev_center_network_connection_resource_test.go:66]: Updated

**Review by ms-zhenhua — APPROVED** [review]
@neil-yechenwei, thank you for your updates. LGTM~

**Review by ms-zhenhua — APPROVED** [review]
@neil-yechenwei, thank you for your updates. LGTM~

File: internal/services/devcenter/dev_center_network_connection_resource.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+						state.DomainName = pointer.From(v)
+					}
+
+					if v := props.DomainUsername; v != nil {
+						state.DomainUsername = pointer.From(v)
+					}
+
+					if v := props.OrganizationUnit; v != nil {
+						state.OrganizationUnit = pointer.From(v)
+					}
```
Comment (stephybun) [inline internal/services/devcenter/dev_center_network_connection_resource.go:216]: `pointer.From` can handle the nil check for us so these can be simplified
```suggestion
	
					state.SubnetId = pointer.From(props.SubnetId)
					state.DomainName = pointer.From(props.DomainName)
					state.DomainUsername = pointer.From(props.DomainUsername)
					state.OrganizationUnit = pointer.From(props.OrganizationUnit)
```

File: internal/services/devcenter/dev_center_network_connection_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+resource "azurerm_virtual_network" "test" {
+  name                = "acctest-vnet-%d"
+  address_space       = ["10.0.0.0/16"]
+  location            = azurerm_resource_group.test.location
+  resource_group_name = azurerm_resource_group.test.name
+
+  lifecycle {
+    ignore_changes = [subnet]
+  }
```
Comment (stephybun) [inline internal/services/devcenter/dev_center_network_connection_resource_test.go:232]: This change has been reverted so we should remove this 
```suggestion
```

File: website/docs/r/dev_center_network_connection.html.markdown [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+resource "azurerm_virtual_network" "example" {
+  name                = "example-vnet"
+  address_space       = ["10.0.0.0/16"]
+  location            = azurerm_resource_group.example.location
+  resource_group_name = azurerm_resource_group.example.name
+
+  lifecycle {
+    ignore_changes = [subnet]
+  }
```
Comment (stephybun) [inline website/docs/r/dev_center_network_connection.html.markdown:29]:
```suggestion
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @neil-yechenwei. Overall this is looking good, once you've fixed up the few minor comments then this will be good to go!

File: website/docs/r/dev_center_network_connection.html.markdown (hunk 4 above)
Comment (neil-yechenwei) [reply website/docs/r/dev_center_network_connection.html.markdown]: Updated

File: internal/services/devcenter/dev_center_network_connection_resource_test.go (hunk 3 above)
Comment (neil-yechenwei) [reply internal/services/devcenter/dev_center_network_connection_resource_test.go]: Updated

File: internal/services/devcenter/dev_center_network_connection_resource.go (hunk 2 above)
Comment (neil-yechenwei) [reply internal/services/devcenter/dev_center_network_connection_resource.go]: Updated

**Comment by neil-yechenwei** [comment]
@stephybun , thanks for the comments. I updated PR. Below is the latest test result I just now triggered.

image
image
image
image

**Review by stephybun — APPROVED** [review]
Thanks @neil-yechenwei LGTM 💯

**Pull request merged**